rcl_action/include/rcl_action/goal_handle.h

```diff
@@ -0,0 +1,347 @@
+ * // ... error handling
+ * rcl_action_goal_handle_t goal_handle = rcl_action_get_zero_initialized_goal_handle();
+ * ret = rcl_action_goal_handle_init(&goal_handle, &action_server, ts);
+ * // ... error handling, and on shutdown do finalization:
+ * ret = rcl_action_goal_handle_fini(&goal_handle, &action_server);
+ * // ... error handling for rcl_goal_handle_fini()
+ * ret = rcl_action_server_fini(&action_server, &node);
+ * // ... error handling for rcl_action_server_fini()
+ * // ... finalize and error handling for node
+ * ```
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[out] goal_handle preallocated, zero-initialized, goal handle structure
+ *   to be initialized
+ * \param[in] action_server valid rcl action server
+ * \param[in] type_support type support object for the action's type
+ * \return `RCL_RET_OK` if goal_handle was initialized successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_GOAL_HANDLE_INVALID` if the goal handle is invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_BAD_ALLOC` if allocating memory failed, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_goal_handle_init(
+  rcl_action_goal_handle_t * goal_handle,
+  const rcl_action_server_t * action_server,
+  const rosidl_action_type_support_t * type_support);
+
+/// Finalize a rcl_action_goal_handle_t.
+/**
+ * After calling, rcl_action_goal_handle_t will no longer be valid and
+ * rcl_action_server_t will no longer track the goal associated with the goal handle.
+ *
+ * After calling, calls to rcl_action_publish_feedback(), rcl_action_publish_status(),
+ * rcl_action_update_goal_state(), rcl_action_goal_handle_get_status(),
+ * rcl_action_goal_handle_is_active(), rcl_action_goal_handle_get_message(), and
+ * rcl_action_goal_handle_get_info() will fail when using this goal handle.
+ *
+ * However, the given action server is still valid.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] goal_handle struct to be deinitialized
+ * \param[in] action_server used to create the goal handle
+ * \return `RCL_RET_OK` if the goal handle was deinitialized successfully, or
+ * \return `RCL_RET_ACTION_GOAL_HANDLE_INVALID` if the goal handle is invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_goal_handle_fini(
+  rcl_action_goal_handle_t * goal_handle,
+  const rcl_action_server_t * action_server);
```

Comment:
I would expect `goal_handle` to have a reference to `action_server` already. That avoids accidentally passing the wrong action server in.